Adding layer search functionality

Samweli opened the pull request:
This PR implements layer searching functionality

fixes #30 
fixes #71 
fixes #72

ricardogsilva reviewed (changes requested):
It is a good start :smile_cat: but there's still stuff to adjust. I think we will do one or two more review cycles of this.

As a preamble, please remember to update the PR description when you set it to not be a draft anymore, as the content of the description is no longer valid.

- [x] No feedback is given for invalid GeoNode connections - make use of the `GeonodeClient.error_received` signal and show the user some feedback when a request did not work OK
- [ ] Please do not change the mouse cursor when a request is loading, as that sets an expectation that the GUI is frozen when in reality it isn't - use some other visual indicator (e.g. a progress bar) instead
- [x] Pagination does not work OK for the first page
- [x] There is a lot of wasted space below the search results area - adjust search results widget to take up all possible space
- [x] Re enable the `Search Items` group box (but not the individual widgets inside it) so that we may test how the search results area expands when this area is collapsed/expanded
- [x] Be more consistent with naming: Qt and QGIS want to name everything with `camelCase`. But Python's convention is to use `snake_case` instead, so please use that (e.g. instead of defining a `currentPage` variable, name it `current_page`)
- [ ] For showing the search results, instead of a `QTreeWidget` and multiple `QTreeWidgetItem`s, use a `QScrollArea` combined with the custom widget you had designed earlier (`ui/qgis_geonode_searched_resource.ui`, which by te way you could rename to something terser, like `search_result_widget.ui`). This shall give us more flexibility in presenting the search results (e.g. for showing a thumbnail of each result, etc). Check this nice tutorial on how to use scroll area if needed: https://www.learnpyqt.com/tutorials/qscrollarea/
- [ ] Clear any search results when the user selects a different GeoNode connection from the dropdown at the top
- [ ] Instantiate geonode client by using `GeonodeClient.from_connection_settings`, as that will make sure to include all relevant parameters in the initialization (e.g. you are currently not passing the auth_config)

Samweli commented:
@ricardogsilva I have updated the PR with the latest changes.

> * Please do not change the mouse cursor when a request is loading, as that sets an expectation that the GUI is frozen when in reality it isn't - use some other visual indicator (e.g. a progress bar) instead

I have added a progress message bar, I couldn't get progress bar to work within the given time.



> * For showing the search results, instead of a `QTreeWidget` and multiple `QTreeWidgetItem`s, use a `QScrollArea` combined with the custom widget you had designed earlier (`ui/qgis_geonode_searched_resource.ui`, which by te way you could rename to something terser, like `search_result_widget.ui`)

I have used the `search_result_widget` and QScrollArea, but I'm having some issues in showing the widgets content inside the scroll area, may be there is some setting I'm missing in using the QScrollArea.

ricardogsilva commented on src/qgis_geonode/gui/geonode_source_select_provider.py:
```diff
@@ -2,18 +2,34 @@
+
 from ..gui.connection_dialog import ConnectionDialog
 
+from qgis_geonode.apiclient.api_client import GeonodeClient
+from qgis_geonode.gui.search_result_widget import SearchResultWidget
```
This does not work due to recent changes introduced by #62 - did you test it?

Samweli replied:
the apiclient package remained in my local repo, so it was still working.

ricardogsilva commented on src/qgis_geonode/gui/geonode_source_select_provider.py:
```diff
@@ -2,18 +2,34 @@
 import uuid
 
-from qgis.core import QgsProject
-from qgis.gui import QgsSourceSelectProvider, QgsAbstractDataSourceWidget
+from qgis.core import QgsProject, Qgis
+from qgis.gui import (
+    QgsSourceSelectProvider,
+    QgsAbstractDataSourceWidget,
+    QgsMessageBar,
+)
 
 from qgis.PyQt.uic import loadUiType
 
 from qgis.PyQt.QtGui import QIcon
-from qgis.PyQt.QtWidgets import QMessageBox
+
+from qgis.PyQt.QtCore import Qt
 
 from ..utils import tr
 from ..conf import connections_manager
+
+from qgis.PyQt.QtWidgets import (
+    QMessageBox,
+    QSizePolicy,
+    QWidget,
+    QVBoxLayout
+)
+
 from ..gui.connection_dialog import ConnectionDialog
 
+from qgis_geonode.apiclient.api_client import GeonodeClient
+from qgis_geonode.gui.search_result_widget import SearchResultWidget
```
Please organize your imports so that they go like this:

1. standard library
2. third-party dependencies
3. own code

Also, please use relative imports instead of absolute ones (e.g. prefer using `from .. import api_client` instead of `from qgis_geonode import api_client`)

ricardogsilva replied:
Also, whithin their logical group, please sort imports alphabetically

ricardogsilva reviewed (changes requested):
Some small changes that are blocking me from doing a proper review. Please fix ASAP

ricardogsilva reviewed (changes requested):
After my recent push using the custom search result widget should be working.

Please fix the comments marked with `TODO` and `FIXME` on the files related to this PR and verify if things are working and then re-request a review.

ricardogsilva commented:
@Samweli as stated offline, please fix #71 and #72 too as a part of this PR

ricardogsilva commented on src/qgis_geonode/gui/geonode_source_select_provider.py:
```diff
@@ -110,10 +141,90 @@ def toggle_connection_management_buttons(self):
+    def show_search_error(self, error):
+        self.message_bar.clearWidgets()
+        self.search_btn.setEnabled(True)
+        self.message_bar.pushMessage(
+            tr("Error searching, {}").format(
+                http_error_description(error)
+            ), level=Qgis.Critical
+        )
```
The error that is sent to this function is a [QNetworkReply.NetworkError](https://doc.qt.io/qt-5/qnetworkreply.html#NetworkError-enum), not an HTTP status code. You are mixing these up here, which causes confusion. For example:

![Screenshot from 2021-01-22 10-04-53](https://user-images.githubusercontent.com/732010/105477626-0398da00-5c9a-11eb-9868-92327217c23d.png)

Now I don't see the error code anymore, but I get a message that would indicate success because you are showing the text for HTTP 203 (which is success) as if it were the description of `QNetworkReply.NetworkError.ContentNotFoundError`, which has the code `203`

Please use the name of the Qt error. See here for a possible implementation:

https://stackoverflow.com/a/39677321

Samweli replied:
ooh I completely miscalculated  that!

ricardogsilva commented on src/qgis_geonode/utils.py:
```diff
@@ -8,3 +17,69 @@ def tr(text):
+              'Object moved temporarily -- see URI list'),
+        400: ('Bad Request',
+              'Bad request syntax or unsupported method'),
+        401: ('Unauthorized',
+              'No permission -- see authorization schemes'),
+        402: ('Payment Required',
+              'No payment -- see charging schemes'),
+        403: ('Forbidden',
+              'Request forbidden -- authorization will not help'),
+        404: ('Not Found', 'Nothing matches the given URI'),
+        405: ('Method Not Allowed',
+              'Specified method is invalid for this server.'),
+        406: ('Not Acceptable', 'URI not available in preferred format.'),
+        407: ('Proxy Authentication Required', 'You must authenticate with '
+                                               'this proxy before proceeding.'),
+        408: ('Request Timeout', 'Request timed out; try again later.'),
+        409: ('Conflict', 'Request conflict.'),
+        410: ('Gone',
+              'URI no longer exists and has been permanently removed.'),
+        411: ('Length Required', 'Client must specify Content-Length.'),
+        412: ('Precondition Failed', 'Precondition in headers is false.'),
+        413: ('Request Entity Too Large', 'Entity is too large.'),
+        414: ('Request-URI Too Long', 'URI is too long.'),
+        415: ('Unsupported Media Type', 'Entity body in unsupported format.'),
+        416: ('Requested Range Not Satisfiable',
+              'Cannot satisfy request range.'),
+        417: ('Expectation Failed',
+              'Expect condition could not be satisfied.'),
+        500: ('Internal Server Error', 'Server got itself in trouble'),
+        501: ('Not Implemented',
+              'Server does not support this operation'),
+        502: ('Bad Gateway', 'Invalid responses from another server/proxy.'),
+        503: ('Service Unavailable',
+              'The server cannot process the request due to a high load'),
+        504: ('Gateway Timeout',
+              'The gateway server did not receive a timely response'),
+        505: ('HTTP Version Not Supported', 'Cannot fulfill request.'),
+    }
+
+    return errors[code]
```
as mentioned in the comment above, these should not be here, as we are not dealing with HTTP status codes directly. Please remove.

ricardogsilva reviewed (changes requested):
Please fix the error descriptions, relevant for #72

ricardogsilva reviewed (approved):
All seems to be working. Thanks for the effort!
